### ISSUE TITLE
Explore behavior if `disconnect()` is called while the container is attempting to connect

### ISSUE THREAD

**Issue by scottn12** [microsoft/FluidFramework#9789]
Investigate issues when calling `disconnect()` while a container is already trying to connect.

Related: https://github.com/microsoft/FluidFramework/issues/9787

**Comment by scottn12** [comment]
This issue has been confirmed. If a container is trying to establish connection and `disconnect()` is called, the `disconnect()` call will fail and the container will still try to complete the connection. This is currently due to a check in [ConnectionManager.disconnectFromDeltaStream()](https://github.com/microsoft/FluidFramework/blob/160894b09d74a7ef366b1a9ae27aa090c7d2d5fc/packages/loader/container-loader/src/connectionManager.ts#L544):
```ts
if (this.connection === undefined) {
    return false;
}

assert(!this.pendingConnection, 0x27b /* "reentrancy may result in incorrect behavior" */);
```

To handle this scenario, I propose we add a mechanism to cancel a pending connection. This could be done by modifying the loop inside `ConnectionManager.connectCore()`. We can add a flag, `connectionCancelled` for example, which would be checked at the start of each iteration of the loop. We would also want to check this flag upon a successful connection to ensure that the flag was not set mid-iteration of the loop. This could be added to `ConnectionManager.setupNewSuccessfulConnection()`. See pseudo-code below:

[ConnectionManager.connectCore()](https://github.com/microsoft/FluidFramework/blob/160894b09d74a7ef366b1a9ae27aa090c7d2d5fc/packages/loader/container-loader/src/connectionManager.ts#L455):
```ts
// This loop will keep trying to connect until successful, with a delay between each iteration.
while (connection === undefined) {
    if (this.connectionCancelled) {
        // Exit loop, ensure all variables, listeners, etc are cleaned up, send disconnected event
    }
    ...
}

this.setupNewSuccessfulConnection(connection, requestedMode);
```

[ConnectionManager.setupNewSuccessfulConnection](https://github.com/microsoft/FluidFramework/blob/160894b09d74a7ef366b1a9ae27aa090c7d2d5fc/packages/loader/container-loader/src/connectionManager.ts#L579):
```ts
private setupNewSuccessfulConnection(connection: IDocumentDeltaConnection, requestedMode: ConnectionMode) {
    ...
    if (this.connectionCancelled) {
        // Raise proper events, Log telemetry event and close connection.
        this.disconnectFromDeltaStream("Connection attempt cancelled");
        return;
    }
    ...
}

```

Thoughts, @vladsud @markfields?

**Comment by vladsud** [comment]
Looking at the code, I'd think 0x27b assert is slightly unrelated cause.
1. There is only one place in the code where we set this.connection to a value other than undefined. And that same place resets this.pendingConnection to false couple lines above.
2. Same in reverse - we set this.pendingConnection = true only after checking that this.connection == undefined.

So I do not see how it's possible to have this.connection !== undefined && this.pendingConnection === true at the same time. Basically, there is invariant that (this.connection === undefined || !this.pendingConnection) that should be true through the code (and it breaks for 2 lines of code when we transition states).
Can you please take a closer look how we break this invariant?

**Comment by vladsud** [comment]
I'd prefer to start using standard tools to deal with cancelation.
I.e. AbortSignal, and even better - IProgress.
And start propagating it to connectToDeltaStream() call itself (i.e. change its signature).
In future, I'd love to start propagating info back, i.e. using IProgress.onRetry() as a mechanism to communicate back about 429s, offline and any other retryable errors we are hitting. And find a way to expose such IProgress more globally (as I think we can't avoid it, even though I'd love).

Related:
- https://github.com/microsoft/FluidFramework/issues/5192

**Comment by scottn12** [comment]
> So I do not see how it's possible to have this.connection !== undefined && this.pendingConnection === true at the same time. Basically, there is invariant that (this.connection === undefined || !this.pendingConnection) that should be true through the code (and it breaks for 2 lines of code when we transition states).
Can you please take a closer look how we break this invariant?

It seems that having `this.pendingConnection` is a bit redundant, since it seems to simply indicate if `this.connection` is defined/undefined. I'm not sure if it was originally intended to act as more of a connecitng state indicator. If so, I like Mark's suggestion to consolidate the state tracking variables with something like `connectingStatus`. We could remove `this.pendingConnection` all together, and use `this.connection === undefined` in its place.

>I'd prefer to start using standard tools to deal with cancelation.
I.e. AbortSignal, and even better - IProgress.
And start propagating it to connectToDeltaStream() call itself (i.e. change its signature).
In future, I'd love to start propagating info back, i.e. using IProgress.onRetry() as a mechanism to communicate back about 429s, offline and any other retryable errors we are hitting. And find a way to expose such IProgress more globally (as I think we can't avoid it, even though I'd love).

@vladsud, please confirm if I have the right understanding. We would want to add a private `AbortSignal`/`IProgress` object to the `Container` class, and use it as a prop in `IConnectionArgs`. It would be passed into `Container.connectToDeltaStream()` and eventually pass it down to `ConnectionManager.connectCore()`, where a listener would be setup for the signal. Then, if `Container.disconnect()` is called, we could fire the abort event to cancel the connection attempt (in a similar way to how I had planned above). Hopefully this is what you have in mind, if not please clarify. I have an idea of what the advantages are of this method, but if you don't mind explaining I'd love to hear your perspective/thought process!

**Comment by vladsud** [comment]
I've added it :) https://github.com/microsoft/FluidFramework/pull/8423/files.
It protects against reentrancy - calling connectCore() multiple times - see the check at the top. It's not the same thing as using this.connection - there are 3 states, thus we need an extra tracking. There used to be a promise there that served same purpose, I downgraded it to boolean as I made this function void (it used to return pending connection). It was incremental change; I'd say we are better off to change state tracking to something like
```ts
type connection = undefined | "pending" | IDocumentDeltaConnection.
```
This clearly articulates state machine.
This also allows more complex structure in the future if we want it, like
```ts
type connection = {state: "none"} | {state: "pending", progress: IProgress; } | { state: "connected", connection: IDocumentDeltaConnection; }
```
or something similar

**Comment by vladsud** [comment]
Yes, something like that, but we should think about lifetime management.

Below is thinking aloud (i.e. mostly brainstorming various directions)

In most other APIs I'd prefer IProgress to be passed into the call as an argument (like getBlob() API), but here I can see us returning IProgress from connect() method (with it being either created for first call, or returning existing one if connection is already in progress). That way clients will have a chance to observer in-flight connection, see its progress and even cancel it.
This assumes that most usage comes in the form of "please connect and report back status".

It can't be global (in a sense created once for lifetime of container), as once it's cancelled once, we will need to throw it away and create new object when attempting next connection. So someone needs to manage its lifetime.

We try a model with passing in IProgress. connect() method may create its own abort signal, and daisy-chain any new objects to it. i.e. something like that:
```ts
connect(progress: IProgress) {
   if (this.cancelSignal === undefined) {
      this.cancelSignal = ... // create one;
   }
   progress.cancel.on("abort", () => this.cancelSignal.abort());

   // Code across DeltaManager only uses this.cancelSignal
}
```

Someone needs to experiment what works the best. :)
But the main key here - once we have abort signals in circulation, we should start passing them up the stack into the driver. And eventually make this state machine unroll fully and leave minimal resources (if any) hanging around. I.e. most fetch() calls and I think socket.io allows passing abort signal, so this should release those resources right away, instead of current model where we allow them to hang around for quite a while.

**Comment by markfields** [comment]
I like this direction a lot.  I've seen a need for a progress tracking object during the connection attempt in my investigation of stalled connections with Loop folks.  Can be out of scope at first, but I'd like to see `IProgress` also support some kind of "steps" property that lists which steps have happened, whether they succeeded or failed, what their duration was. Something lightweight like `steps: { stepName: string, status: "inProgress" | "succeeded" | "failed"; durationMs: number }[]` which can be used for logging.

**Comment by markfields** [comment]
@scottn12 - I'd be happy to pair on this with you tomorrow or next week, let me know if you'd like to take a look together.

**Comment by vladsud** [comment]
IProgress is generic thing. We can always go with custom interfaces (hopefully deriving from it) or templatization - IProgress<T> to allow extensibility. That way each workflow can be strongly typed, yet expose unique characteristics of the process.
